Update `CONTRIBUTING.md`

g-votte opened the pull request:
## Motivation
- Currently, the information for contributors is cluttered between `CONTRIBUTING.md` and [Wiki](https://github.com/optuna/optuna/wiki). This PR recompiles those pieces of information referencing them from `CONTRIBUTING.md`.
- Currently, some parts of the section structure of `CONTRIBUTING.md` is asymmetry. For example, `Unit Tests` and `Documentation` are placed in different layer despite that both subsections are guidelines to create pull request. This PR restructures them.

## Description of the changes
- Reference the conventions and policies in Optuna Wiki
- Restructure the sections as:
    - Contribution Guidelines
    - Pull Request Guidelines
        - Setup Optuna
        - Checking the Format, Coding Style, and Typle Hinsts
        - Documentation
        - Unit Tests
        - Continuous Integration and Local Verification
        - Creating a Pull Request
    - Learning Optuna Implementation

toshihikoyanase commented:
@contramundum53 @nzw0301 Could you review this PR, please? If you are busy, please feel free to remove the assignment.

contramundum53 commented on CONTRIBUTING.md:
```diff
@@ -26,15 +26,23 @@ If you feel like giving a hand, here are some ways:
-If you write code, we have some conventions as follows.
 
-- [Guidelines](#guidelines)
+# Pull Request Guidelines
```
```suggestion
## Pull Request Guidelines
```
The first-level header is already used for title.

contramundum53 commented:
Isn't `Pull Request` part of `Contribution`?

contramundum53 commented on CONTRIBUTING.md:
```diff
@@ -26,15 +26,23 @@ If you feel like giving a hand, here are some ways:
+- [Deprecation Policy](https://github.com/optuna/optuna/wiki/Deprecation-policy)
+- [Test Policy](https://github.com/optuna/optuna/wiki/Test-Policy)
 
 ### Setup Optuna
```
```suggestion
## Setup Optuna
```

g-votte replied:
Thanks for pointing this out. After [this change](https://github.com/optuna/optuna/pull/3726#issuecomment-1169485567), "Setup Optuna" is actually the 3rd level. Thus, I'll leave this line as it is.

nzw0301 commented:
I've been working a little bit too long this month, so let me unassign myself once.

g-votte commented:
@contramundum53 

Thanks for your comment. I changed the structure again as:
- Contribution Guidelines
  - Pull Request Guidelines
    - Setup Optuna
    - Checking the Format, Coding Style, and Typle Hinsts
    - Documentation
    - Unit Tests
    - Continuous Integration and Local Verification
    - Creating a Pull Request
  - Learning Optuna Implementation

In addition, I removed the subsections of `title`, `example`,  and `description` in `Creating a Pull Request` because they've been too much nested so that the font size is almost the same as normal text.

contramundum53 reviewed (approved):
LGTM!

HideakiImamura reviewed (approved):
LGTM.
